From 1f99dc1b7771c252cd8e6e75ef5818e446f11c22 Mon Sep 17 00:00:00 2001
From: Michael Olbrich <m.olbrich@pengutronix.de>
Date: Wed, 4 Jul 2018 09:21:00 +0200
Subject: [PATCH 08/45] glupload: handle upload methods with different caps

If a upload method is selected then use it exclusively in transform_caps().
Also, reconfigure if the current caps don't match the current upload
method.

https://bugzilla.gnome.org/show_bug.cgi?id=783521
(cherry picked from commit f349cdccf5e1538f3eb9caa31458b53fdd81671b)
Signed-off-by: Jeffy Chen <jeffy.chen@rock-chips.com>
---
 gst-libs/gst/gl/gstglupload.c | 24 ++++++++++++++++++++++++
 1 file changed, 24 insertions(+)

diff --git a/gst-libs/gst/gl/gstglupload.c b/gst-libs/gst/gl/gstglupload.c
index b53d481f9e..5ec84a8cac 100644
--- a/gst-libs/gst/gl/gstglupload.c
+++ b/gst-libs/gst/gl/gstglupload.c
@@ -1669,6 +1669,21 @@ gst_gl_upload_transform_caps (GstGLUpload * upload, GstGLContext * context,
   GstCaps *result, *tmp;
   gint i;
 
+  if (upload->priv->method) {
+    tmp = upload->priv->method->transform_caps (upload->priv->method_impl,
+        context, direction, caps);
+    if (filter) {
+      result = gst_caps_intersect_full (filter, tmp, GST_CAPS_INTERSECT_FIRST);
+      gst_caps_unref (tmp);
+    } else {
+      result = tmp;
+    }
+    if (!gst_caps_is_empty (result))
+      return result;
+    else
+      gst_caps_unref (result);
+  }
+
   tmp = gst_caps_new_empty ();
 
   for (i = 0; i < G_N_ELEMENTS (upload_methods); i++) {
@@ -1872,6 +1887,15 @@ restart:
     }
     goto restart;
   } else if (ret == GST_GL_UPLOAD_DONE || ret == GST_GL_UPLOAD_RECONFIGURE) {
+    if (last_impl != upload->priv->method_impl) {
+      GstCaps *caps = gst_gl_upload_transform_caps (upload, upload->context,
+          GST_PAD_SINK, upload->priv->in_caps, NULL);
+      if (!gst_caps_is_equal (upload->priv->out_caps, caps)) {
+        gst_buffer_replace (&outbuf, NULL);
+        ret = GST_GL_UPLOAD_RECONFIGURE;
+      }
+      gst_caps_unref (caps);
+    }
     /* we are done */
   } else {
     upload->priv->method_impl = NULL;
-- 
2.11.0

